node/src/main/scala/com/wavesplatform/mining/BlockChallenger.scala

```diff
@@ -0,0 +1,277 @@
+    pos: PoSSelector,
+    minerScheduler: Scheduler,
+    appendBlock: Block => Task[Either[ValidationError, Option[BigInt]]]
+) extends BlockChallenger
+    with ScorexLogging {
+
+  private val processingTxs: ConcurrentHashMap[ByteStr, Transaction] = new ConcurrentHashMap()
+
+  def challengeBlock(block: Block, ch: Channel, prevStateHash: ByteStr, blockReward: Option[Long]): Task[Unit] = {
+    log.debug(s"Challenging block $block")
+
+    (for {
+      _ <- EitherT.liftF(Task(processingTxs.putAll(block.transactionData.map(tx => tx.id() -> tx).toMap.asJava)))
+      challengingBlock <- EitherT(
+        createChallengingBlock(block, block.header.stateHash, block.signature, block.transactionData, prevStateHash, blockReward)
+      )
+      _ <- EitherT(appendBlock(challengingBlock).asyncBoundary)
+      _ <- EitherT.liftF[Task, ValidationError, Unit](Task(processingTxs.clear()))
+    } yield {
+      log.debug(s"Successfully challenged $block with $challengingBlock")
+      allChannels.broadcast(BlockForged(challengingBlock), Some(ch))
+    }).fold(
+      err => {
+        processingTxs.clear()
+        log.debug(s"Could not challenge $block: $err")
+      },
+      identity
+    )
+  }
+
+  def challengeMicroblock(md: MicroblockData, ch: Channel, prevStateHash: ByteStr, blockReward: Option[Long]): Task[Unit] = {
+    val idStr = md.invOpt.map(_.totalBlockId.toString).getOrElse(s"(sig=${md.microBlock.totalResBlockSig})")
+    log.debug(s"Challenging microblock $idStr")
+
+    (for {
+      discarded <- EitherT(Task(blockchainUpdater.removeAfter(blockchainUpdater.lastBlockHeader.get.header.reference)))
+      block     <- EitherT(Task(discarded.headOption.map(_._1).toRight(GenericError("Liquid block wasn't discarded"))))
+      _ <- EitherT.liftF(Task(processingTxs.putAll((block.transactionData ++ md.microBlock.transactionData).map(tx => tx.id() -> tx).toMap.asJava)))
+      challengingBlock <- EitherT(
+        createChallengingBlock(
+          block,
+          md.microBlock.stateHash,
+          md.microBlock.totalResBlockSig,
+          block.transactionData ++ md.microBlock.transactionData,
+          prevStateHash,
+          blockReward
+        )
+      )
+      _ <- EitherT(appendBlock(challengingBlock).asyncBoundary)
+      _ <- EitherT.liftF[Task, ValidationError, Unit](Task(processingTxs.clear()))
+    } yield {
+      log.debug(s"Successfully challenged microblock $idStr with $challengingBlock")
+      allChannels.broadcast(BlockForged(challengingBlock), Some(ch))
+    }).fold(
+      err => {
+        processingTxs.clear()
+        log.debug(s"Could not challenge microblock $idStr: $err")
+      },
+      identity
+    )
+  }
+
+  def pickBestAccount(accounts: Seq[(SeedKeyPair, Long)]): Either[GenericError, (SeedKeyPair, Long)] =
+    accounts.minByOption(_._2).toRight(GenericError("No suitable account in wallet"))
+
+  def getChallengingAccounts(challengedMiner: Address): Either[ValidationError, Seq[(SeedKeyPair, Long)]] =
```

Comment:
It seems, `pickBestAccount` and `getChallengingAccounts` always are used both. Probably, it'd better to make `def pickBestChallengingAccount(challengedMiner: Address): Either[GenericError, (SeedKeyPair, Long)]` ?

Reply:
It was done for testing purposes: https://github.com/wavesplatform/Waves/pull/3859/files#diff-0696e3ee17bb75c4e1bc89d88cc58f915014d9af7d321678d6eb7ed3c0cd1f18R178-R180